kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/assertions/RetryTest.kt

```diff
@@ -27,100 +41,185 @@ class RetryTest : StringSpec() {
+         retryTester.calledAtTimeInstance shouldHaveSize 2
       }
-   }
 
-   private class TestClass(private val readyAfter: Int) {
-      var calledAtTimeInstance = listOf<Long>()
-      var times = 0
-      fun isReady(): Boolean {
-         calledAtTimeInstance = calledAtTimeInstance.plus(System.currentTimeMillis())
-         times += 1
-         return readyAfter == times
+      "when maxRetry is negative, expect zero invocations" {
+         val retryTester = retryTester(4)
+         val retryException = shouldThrow<AssertionError> {
+            retry(maxRetry = -1, timeout = 500.milliseconds) {
+               retryTester.isReady() shouldBe true
+            }
+         }
+         retryException shouldHaveMessage "Test failed after 0s; attempted 0 times"
+         retryTester.calledAtTimeInstance.shouldBeEmpty()
       }
 
-      fun throwUnexpectedException() {
-         times += 1
-         throw NullPointerException("")
+      "when multiplier is negative, expect ..." { // TODO
+         val retryTester = retryTester(4)
+         retry(maxRetry = 5, timeout = 500.milliseconds, delay = 100.milliseconds, multiplier = -1) {
+            delay(100)
+            retryTester.isReady() shouldBe true
+         }
+         retryTester.calledAtTimeInstance.shouldContainExactly(
+            100.milliseconds,
+            300.milliseconds,
+            400.milliseconds,
+            600.milliseconds,
+         )
+      }
+
+      "when delay is negative, expect ..." { // TODO
+         val retryTester = retryTester(4)
+         shouldNotThrow<Exception> {
+            retry(maxRetry = 5, timeout = 500.milliseconds, delay = (-100).milliseconds) {
+               delay(100)
+               retryTester.isReady() shouldBe true
+            }
+         }
+         retryTester.calledAtTimeInstance.shouldContainExactly(
+            100.milliseconds,
+            200.milliseconds,
+            300.milliseconds,
+            400.milliseconds,
+         )
+      }
+
+      "when timeout is negative, expect zero invocations" {
+         val retryTester = retryTester(4)
+         val retryException = shouldThrow<AssertionError> {
+            retry(maxRetry = 5, timeout = (-500).milliseconds) {
+               retryTester.isReady() shouldBe true
+            }
+         }
+         retryException shouldHaveMessage "Test failed after 0s; attempted 0 times"
+         retryTester.calledAtTimeInstance.shouldBeEmpty()
       }
    }
 
-   private suspend fun runSafely(block: suspend () -> Unit) {
-      try {
-         block()
-      } catch (assertionError: AssertionError) {
-         // Eating assertion error
-      } catch (exception: Exception) {
-         // Eating other exception
+   private class RetryTester(
+      private val readyAfter: Int,
+      private val timeMark: TimeMark,
+   ) {
+      val calledAtTimeInstance = mutableListOf<Duration>()
+
+      fun isReady(): Boolean {
+         calledAtTimeInstance += timeMark.elapsedNow()
```

Comment:
I find it confusing if a function returns a value _and_ has a side effect as there is no indication of such behavior at call sites. This one is more like `advance()` plus `isReady()`. If splitting is not desirable, I'd consider something like `advance()` returning a result class like `RetryTester.Result(val isReady: Boolean)` a good compromise.

Just my thoughs. Leave it up to you to decide.

Reply:
Yeah, I agree. I don't find the RetryTester class to be particularly clear. But I think it's a useful util, and I didn't want to completely re-write everything, so I just amended it.

I think it could probably be replaced with some `testRetry`/`testRetryFails` functions.